### ISSUE TITLE
Usage as map key

### ISSUE THREAD

**Issue by mmank** [shopspring/decimal#40]
As it stands, Decimal does not work as map key. This is because it uses `value *big.Int`. Why not just `value big.Int` ? Is there a real performance reason?

**Comment by aaronbee** [comment]
Even if `value` were to be changed to a concrete `big.Int`, you would not be able to use `Decimal` as a map key. `big.Int` contains a `nat` which is a `[]Word` (`uint`). Any struct that has a slice will cause a panic when you try to use it as a map key.

**Comment by haifenghuang** [comment]
Maybe you could use [hashstructure](https://github.com/mitchellh/hashstructure).

**Comment by dimonomid** [comment]
Yeah it's really annoying that a `Decimal` can't be used as a map key. In my app I need it a lot.

My benchmarks show that just converting a `Decimal` to string (by calling `String()`) is ~4 times faster than using hashstructure, so I just use `String()`.

Still, `String()` is way too slow if we need to deal with lots of `Decimal`s. The workaround I'm thinking of is to cache string value of a `Decimal` right in there. If `Decimal` was created from a string, it'd be populated right away; otherwise it would be calculated lazily.

I'm sure it won't be accepted into the decimal repo though, I'll probably just keep using my wrapper which caches strings like that.

**Comment by mwoss** [comment]
As @aaronbee explained it not possible for us to easily create hashable decimal objects with the current implementation. 
We will try to rethink the whole concept of hashability of decimal objects soon and let you know.

**Comment by weaming** [comment]
Any updates?
